### PR TITLE
[Graphbolt]Unique and compact OP

### PR DESCRIPTION
**Pull request by peizhou001**
## Description
Change unique and compact op.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: graphbolt/src/concurrent_id_hash_map.cc [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+inline size_t GetMapSize(size_t num) {
+  size_t capacity = 1;
+  return capacity << static_cast<size_t>(1 + std::log2(num * 3));
+}
+}  // namespace
+
+namespace graphbolt {
+namespace sampling {
+
+template <typename IdType>
```
Comment (mfbalin) [inline graphbolt/src/concurrent_id_hash_map.cc:31]: Why not use: https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange

File: graphbolt/src/unique_and_compact.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * @brief Removes duplicate elements from the concatenated 'unique_dst_ids' and
+ * 'src_ids' tensors and applies the uniqueness information to compact both
+ * source and destination tensor lists.
+ *
+ * The function performs two main operations:
+ *   1. Unique Operation: Concatenates 'unique_dst_ids' and 'src_ids' tensors
+ * and extracts unique elements, ensuring no duplicates. The resulting tensor
+ * represents all unique elements in both 'src_ids' and 'dst_ids', and the
+ * indices correspond to the compacted IDs of the corresponding elements.
```
Comment (frozenbugs) [inline graphbolt/src/unique_and_compact.h:25]: 1. Unique Operation: unique(concat(unique_dst_ids, src_ids)), in which the unique operator will guarantee the unique_dst_ids are at the head of the result tensor.

File: graphbolt/src/unique_and_compact.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ *   TensorList src_tensors = {src_tensor1, src_tensor2, src_tensor3};
+ *   TensorList dst_tensors = {dst_tensor1, dst_tensor2, dst_tensor3};
+ *   torch::Tensor unique_dst_ids = torch::unique(torch::cat(dst_tensors));
+ *   auto result = Unique_and_compact(src_tensors, dst_tensors, unique_dst_ids);
+ *   torch::Tensor unique_result = std::get<0>(result);
+ *   TensorList unique_src_tensors = std::get<1>(result);
+ *   TensorList unique_dst_tensors = std::get<2>(result);
+ */
+std::tuple<torch::Tensor, TensorList, TensorList> Unique_and_compact(
+    const TensorList& src_ids, const TensorList& dst_ids,
```
Comment (frozenbugs) [inline graphbolt/src/unique_and_compact.h:54]: use tensor instead of tensor list.

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 4]
```diff
@@ -52,44 +53,54 @@ def unique_and_compact_node_pairs(
...
+        src_nodes[etype[0]].append(src_node)
+
+    ntypes = set(dst_nodes.keys()) | set(src_nodes.keys())
+
+    unique_nodes = {}
+    compacted_src = defaultdict(list)
+    compacted_dst = defaultdict(list)
+    for ntype in ntypes:
+        src, dst = src_nodes[ntype], dst_nodes[ntype]
+        # When 'unique_dst_nodes' is empty, 'dst' must also be empty and 'src' must have value.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:85]: line too long.

File: graphbolt/src/unique_and_compact.cc [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+        }
+        for (auto id : dst_ids) {
+          compacted_dst_ids.emplace_back(id_map.MapIds(id));
+        }
+      }));
+  return std::tuple(unique_ids, compacted_src_ids, compacted_dst_ids);
+}
+}  // namespace sampling
+}  // namespace graphbolt
+#endif  // GRAPHBOLT_UNIQUE_AND_COMPACT_H_
```
Comment (frozenbugs) [inline graphbolt/src/unique_and_compact.cc:42]: new line

File: graphbolt/src/unique_and_compact.cc [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+#ifndef GRAPHBOLT_UNIQUE_AND_COMPACT_H_
+#define GRAPHBOLT_UNIQUE_AND_COMPACT_H_
+
+#include "./unique_and_compact.h"
+
+#include "./concurrent_id_hash_map.h"
+
+namespace graphbolt {
+namespace sampling {
+std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> Unique_and_compact(
```
Comment (czkkkkkk) [inline graphbolt/src/unique_and_compact.cc:16]:
```suggestion
std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> UniqueAndCompact(
```

File: graphbolt/src/python_binding.cc [hunk 7]
```diff
@@ -7,6 +7,8 @@
 #include <graphbolt/csc_sampling_graph.h>
 #include <graphbolt/serialize.h>
 
+#include "unique_and_compact.h"
```
Comment (czkkkkkk) [inline graphbolt/src/python_binding.cc:10]:
```suggestion
#include "./unique_and_compact.h"
```

File: graphbolt/src/unique_and_compact.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ *
+ * @file unique_and_compact.h
+ * @brief Unique and compact op.
+ */
+#ifndef GRAPHBOLT_SHM_UTILS_H_
```
Comment (czkkkkkk) [inline graphbolt/src/unique_and_compact.h:7]: Header name.

File: graphbolt/src/unique_and_compact.h [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+ * @brief Removes duplicate elements from the concatenated 'unique_dst_ids' and
+ * 'src_ids' tensor and applies the uniqueness information to compact both
+ * source and destination tensors.
+ *
+ * The function performs two main operations:
+ *   1. Unique Operation: unique(concat(unique_dst_ids, src_ids)), in which the
+ * unique operator will guarantee the unique_dst_ids are at the head of the
+ * result tensor.
+ *   2. Compact Operation: Utilizes the reverse mapping derived from the unique
+ * operation to transform 'src_ids' and 'dst_ids' into condensed IDs, mapping
```
Comment (czkkkkkk) [inline graphbolt/src/unique_and_compact.h:24]: We have several terms when dealing with ID mapping, e.g., condensed ID, original ID, inverse ID, and relabeled ID. @frozenbugs Do we have a standard for this kind of names?

File: graphbolt/src/unique_and_compact.cc [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    const torch::Tensor& src_ids, const torch::Tensor& dst_ids,
+    const torch::Tensor unique_dst_ids) {
+  torch::Tensor compacted_src_ids;
+  torch::Tensor compacted_dst_ids;
+  torch::Tensor unique_ids;
+  auto num_dst = unique_dst_ids.size(0);
+  torch::Tensor ids = torch::cat({unique_dst_ids, src_ids});
```
Comment (czkkkkkk) [inline graphbolt/src/unique_and_compact.cc:23]: What would happen if `dst_ids` is not a subset of `unique_dst_ids`?

File: graphbolt/src/unique_and_compact.h (hunk 8 above)
Comment (frozenbugs) [reply graphbolt/src/unique_and_compact.h:7]: GRAPHBOLT_UNIQUE_AND_COMPACT_H_

File: graphbolt/src/unique_and_compact.cc [hunk 11]
```diff
@@ -0,0 +1,34 @@
+/**
+ *  Copyright (c) 2023 by Contributors
+ *
+ * @file unique_and_compact.cc
+ * @brief Unique and compact op.
+ */
+#ifndef GRAPHBOLT_UNIQUE_AND_COMPACT_H_
```
Comment (frozenbugs) [inline graphbolt/src/unique_and_compact.cc:7]: not needed.

File: graphbolt/src/unique_and_compact.h (hunk 9 above)
Comment (frozenbugs) [reply graphbolt/src/unique_and_compact.h:24]: Let's finalize them and doc it here.
https://docs.google.com/spreadsheets/d/1dv41M-Th20e1ppKvX9wUQ8MFFJKRTiRM6PalYAx9mlk/edit#gid=25977840

File: graphbolt/src/unique_and_compact.cc (hunk 10 above)
Comment (frozenbugs) [reply graphbolt/src/unique_and_compact.cc:23]: Will MapIds fail?

Comment (peizhou001) [reply graphbolt/src/unique_and_compact.cc:23]: No, it return -1 for unfounded ids.

File: graphbolt/src/unique_and_compact.h (hunk 9 above)
Comment (peizhou001) [reply graphbolt/src/unique_and_compact.h:24]: As the function is called unique_and_compact, use `compacted_ids `for uniform.

File: graphbolt/src/unique_and_compact.cc (hunk 10 above)
Comment (czkkkkkk) [reply graphbolt/src/unique_and_compact.cc:21]: Is this an expected behavior? Or we need to raise a runtime error?

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 12]
```diff
@@ -10,7 +10,8 @@ def unique_and_compact_node_pairs(
     node_pairs: Union[
         Tuple[torch.Tensor, torch.Tensor],
         Dict[Tuple[str, str, str], Tuple[torch.Tensor, torch.Tensor]],
-    ]
+    ],
+    unique_dst_nodes=None,
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:14]: Update python doc?

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 13]
```diff
@@ -52,44 +53,55 @@ def unique_and_compact_node_pairs(
...
-    for ntype, nodes in nodes_dict.items():
-        collected_nodes = torch.cat(nodes)
-        # Compact and find unique nodes.
-        unique_nodes, inverse_indices = torch.unique(
-            collected_nodes,
-            return_inverse=True,
-        )
-        unique_nodes_dict[ntype] = unique_nodes
-        inverse_indices_dict[ntype] = inverse_indices
+    # Collect all source and destination nodes.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:61]:
```suggestion
    # Collect all source and destination nodes for each node type.
```

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 12 above)
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:14]: What's the type of the unique_dst_nodes?

File: tests/python/pytorch/graphbolt/test_graphbolt_utils.py [hunk 14]
```diff
@@ -46,27 +32,35 @@ def test_unique_and_compact_node_pairs_hetero():
...
+    u, v = unique_nodes[u], unique_nodes[v]
+    expected_u, expected_v = node_pairs
+    unique_v = torch.unique(expected_v)
     assert torch.equal(u, expected_u)
     assert torch.equal(v, expected_v)
-    assert torch.equal(unique_nodes, expected_unique_N)
+    assert torch.equal(unique_nodes[: unique_v.size(0)], unique_v)
+
+
+test_unique_and_compact_node_pairs_hetero()
```
Comment (frozenbugs) [inline tests/python/pytorch/graphbolt/test_graphbolt_utils.py:66]: remove unnecessary code.

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 15]
```diff
@@ -52,44 +53,55 @@ def unique_and_compact_node_pairs(
...
+    src_nodes = defaultdict(list)
+    dst_nodes = defaultdict(list)
+    for etype, (src_node, dst_node) in node_pairs.items():
+        src_nodes[etype[0]].append(src_node)
+        dst_nodes[etype[2]].append(dst_node)
+    src_nodes = {ntype: torch.cat(nodes) for ntype, nodes in src_nodes.items()}
+    dst_nodes = {ntype: torch.cat(nodes) for ntype, nodes in dst_nodes.items()}
+    # Compute unique destination nodes if not provided.
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:69]: nit: Is it slightly better to do this in C++ code?

File: graphbolt/src/unique_and_compact.h [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+ *   1. Unique Operation: 'unique(concat(unique_dst_ids, src_ids))', in which
+ * the unique operator will guarantee the 'unique_dst_ids' are at the head of
+ * the result tensor.
+ *   2. Compact Operation: Utilizes the reverse mapping derived from the unique
+ * operation to transform 'src_ids' and 'dst_ids' into compacted IDs.
+ *
+ * @param src_ids         A tensor containing source IDs.
+ * @param dst_ids         A tensor containing destination IDs.
+ * @param unique_dst_ids  A tensor containing unique destination IDs, which is
+ * exactly all the unique elements in 'dst_ids'.
```
Comment (frozenbugs) [inline graphbolt/src/unique_and_compact.h:29]:
```suggestion
 * @param unique_dst_ids  A tensor containing unique destination IDs, which is
 *                        exactly all the unique elements in 'dst_ids'.
```

File: graphbolt/src/unique_and_compact.h [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+ * - The tensor corresponding to the 'src_ids' tensor, where the entries are
+ * mapped to compacted IDs.
+ * - The tensor corresponding to the 'dst_ids' tensor, where the entries are
+ * mapped to compacted IDs.
+ *
+ * @example
+ *   torch::Tensor src_ids = src
+ *   torch::Tensor dst_ids = dst
+ *   torch::Tensor unique_dst_ids = torch::unique(dst);
+ *   auto result = Unique_and_compact(src_ids, dst_ids, unique_dst_ids);
```
Comment (frozenbugs) [inline graphbolt/src/unique_and_compact.h:44]:
```suggestion
 *   auto result = UniqueAndCompact(src_ids, dst_ids, unique_dst_ids);
```

**Comment by frozenbugs** [comment]
Overall LGTM, added some nit comments.

File: graphbolt/src/unique_and_compact.cc (hunk 10 above)
Comment (frozenbugs) [reply graphbolt/src/unique_and_compact.cc:21]: Good point, maybe we should raise an error here https://github.com/dmlc/dgl/blob/7e4c226de460e0b3ac82952ea7a075c0256f8b5a/graphbolt/src/concurrent_id_hash_map.cc#L163C21-L163C21.
If so, let's draft a separated PR for it, and also add test.

File: graphbolt/src/unique_and_compact.h (hunk 9 above)
Comment (frozenbugs) [reply graphbolt/src/unique_and_compact.h]: SG, added to the doc.

File: graphbolt/src/unique_and_compact.cc (hunk 10 above)
Comment (peizhou001) [reply graphbolt/src/unique_and_compact.cc:21]: Sure, I'll draft a separate PR and add test in this PR.

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 12 above)
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:14]: Added

File: python/dgl/graphbolt/utils/sample_utils.py (hunk 15 above)
Comment (peizhou001) [reply python/dgl/graphbolt/utils/sample_utils.py:69]: This preprocess is all about tensor op, I want to keep C++ code small and preformance concentrated.

**Review by czkkkkkk — APPROVED** [review]

File: python/dgl/graphbolt/utils/sample_utils.py [hunk 18]
```diff
@@ -26,7 +30,12 @@ def unique_and_compact_node_pairs(
         - If `node_pairs` is a dictionary: The keys should be edge type and
         the values should be corresponding node pairs. And IDs inside are
         heterogeneous ids.
-
+    unique_dst_nodes: torch.Tensor or Dict[str, torch.Tensor]
+        Unique nodes of all destination nodes in the node pairs.
+        - If `unique_dst_nodes` is a tensor: It means the graph is homogeneous.
+        - If `node_pairs` is a dictionary: The keys are node type and the
+        values are corresponding nodes. And IDs inside are heterogeneous ids.
+
```
Comment (frozenbugs) [inline python/dgl/graphbolt/utils/sample_utils.py:38]:
```suggestion

```

**Review by frozenbugs — APPROVED** [review]

**Pull request merged**